proposals/lip-0054.md

```diff
@@ -301,120 +289,336 @@ messageRecoveryParams = {
+    ccmID = sha256(encode(crossChainMessageSchema, ccm))
+
+    ccm.status = CCM_STATUS_CODE_RECOVERED
+    ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID
+
+    try:
+        # Modules can verify the CCM.
+        # The Token module verifies the escrowed balance in the CCM sending chain for the message fee.
+        for each module mdl for which verifyCrossChainMessage exists:
+            mdl.verifyCrossChainMessage(trs, ccm)
+    except:
+        emitEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "result": CCM_PROCESSED_RESULT_DISCARDED, "code": CCM_PROCESSED_CODE_INVALID_CCM_VERIFY_CCM_EXCEPTION},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
+        return
+
+    # Create a state snapshot.
+    baseSnapshot = snapshot of the current state
+    try:  
+        # Call the beforeCrossChainMessageForwarding functions from other modules.
+        # For example, the Token module transfers the fee from escrow to escrow.
+        for each module mdl for which beforeCrossChainMessageForwarding exists:
+            mdl.beforeCrossChainMessageForwarding(trs, ccm)
+    except:
+        revert state to baseSnapshot
+        emitEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_PROCESSED,
+            data = {"ccmID": ccmID, "result": CCM_PROCESSED_RESULT_DISCARDED, "code": CCM_PROCESSED_CODE_INVALID_CCM_BEFORE_CCC_FORWARDING_EXCEPTION},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
+        )
+        return
+
+    addToOutbox(ccm.receivingChainID, ccm)
+    # Emit CCM forwarded event.
+    # Recalculate CCM ID. This differs because of new status, sending, and receiving chains.
+    # We use the updated ID as this CCM is appended to the outbox.
+    recoveredCCMID = sha256(encode(crossChainMessageSchema, ccm))
+    emitEvent(
+        module = MODULE_NAME_INTEROPERABILITY,
+        name = EVENT_NAME_CCM_PROCESSED,
+        data = {"ccmID": recoveredCCMID, "result": CCM_PROCESSED_RESULT_FORWARDED, "code": CCM_PROCESSED_CODE_SUCCESS},
+        topics = [ccm.sendingChainID, ccm.receivingChainID]
+    )
+```
+### Liveness Termination Command
+
+This command is used to terminate a sidechain that violated the liveness condition. As such, it can only be submitted on the mainchain.
+
+Transactions executing this command have:
+
+* `module = MODULE_NAME_INTEROPERABILITY`,
+* `command = COMMAND_LIVENESS_TERMINATION`.
+
+#### Parameters
+
+```java
+livenessTerminationParams = {
+    "type": "object",
+    "required": ["chainID"],
+    "properties": {
+        "chainID": {
+            "dataType": "bytes",
+            "length": CHAIN_ID_LENGTH,
+            "fieldNumber": 1
+        }
+}
 ```
 
-#### Message Recovery Command Execution
+#### Verification
+
+```python
+def verify(trs: Transaction) -> None:
```

Comment:
You don't verify explicitly that we are on the mainchain, as you say in the description of the command. Of course, `isLive` function will return `True` on the sidechain, but I would be more precise and add a check `if OWN_CHAIN_ID != CHAIN_ID_MAINCHAIN: raise Exception()`.

Reply:
My point here is that this command would only exist in the mainchain interoperability module (so this check is unnecessary)